browser/widevine/widevine_utils.cc

```diff
@@ -121,9 +121,100 @@ void InstallBundleOrRestartBrowser() {
                                    true);
   }
 }
+
+void SetWidevineInstalledVersion(const std::string& version) {
+  GetLocalState()->SetString(kWidevineInstalledVersion, version);
+}
+
+std::string GetWidevineInstalledVersion() {
+  return GetLocalState()->GetString(kWidevineInstalledVersion);
+}
+#endif
+
+#if BUILDFLAG(ENABLE_WIDEVINE_CDM_COMPONENT) || BUILDFLAG(BUNDLE_WIDEVINE_CDM)
+int GetWidevinePermissionRequestTextFrangmentResourceId() {
+  int message_id = -1;
+
+#if BUILDFLAG(ENABLE_WIDEVINE_CDM_COMPONENT)
+  message_id = IDS_WIDEVINE_PERMISSION_REQUEST_TEXT_FRAGMENT;
 #endif
 
+#if BUILDFLAG(BUNDLE_WIDEVINE_CDM)
+  auto* manager = g_brave_browser_process->brave_widevine_bundle_manager();
+  message_id = manager->GetWidevinePermissionRequestTextFragment();
+#endif
+
+  DCHECK_NE(message_id, -1);
+  return message_id;
+}
+
+void RequestWidevinePermission(content::WebContents* web_contents) {
+  PermissionRequestManager::FromWebContents(web_contents)->AddRequest(
+      new WidevinePermissionRequest(web_contents));
+}
 void DontAskWidevineInstall(content::WebContents* web_contents, bool dont_ask) {
   Profile* profile = static_cast<Profile*>(web_contents->GetBrowserContext());
   profile->GetPrefs()->SetBoolean(kAskWidevineInstall, !dont_ask);
 }
+
+void RegisterWidevineProfilePrefsForMigration(
+    user_prefs::PrefRegistrySyncable* registry) {
+  registry->RegisterBooleanPref(kWidevineOptedIn, false);
+
+#if BUILDFLAG(BUNDLE_WIDEVINE_CDM)
+  registry->RegisterStringPref(
+      kWidevineInstalledVersion,
+      BraveWidevineBundleManager::kWidevineInvalidVersion);
+#endif
+}
+
+void RegisterWidevineLocalstatePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterBooleanPref(kWidevineOptedIn, false);
+
+#if BUILDFLAG(BUNDLE_WIDEVINE_CDM)
+  registry->RegisterStringPref(
+      kWidevineInstalledVersion,
+      BraveWidevineBundleManager::kWidevineInvalidVersion);
+#endif
+}
+
+bool IsWidevineOptedIn() {
+  return GetLocalState()->GetBoolean(kWidevineOptedIn);
+}
+
+void SetWidevineOptedIn(bool opted_in) {
+  GetLocalState()->SetBoolean(kWidevineOptedIn, opted_in);
+}
+
+void MigrateWidevinePrefs(Profile* profile) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+
+  if (g_is_migration_checked) {
```

Comment:
`g_is_migration_checked` for primary profile check is not accurate. user can create a second profile and relaunch with second profile window focused.
So second profile will run `MigrateWidevinePrefs` first and then the primary profile.
Maybe we can check `chrome::kInitialProfile` instead, although it might not work when `Default` profile get removed but it work for the most cases.

Reply:
I assumed firstly initialized profile is user's primary profile in most cases. If user doesn't use initial profile(Default) for some time, initial profile(Default) might not be an user's primary profile.
Also, as you mentioned, checking `chrome:kInitialProfile` (Default) doesn't work when user deleted it. `Default` could be recreated later but it's clean profile.

I changed migration logic w/o `g_is_migration_checked`.
Instead of using it,  `IsDefaultValue()` from local state is used. With this, we can easily check migration is already done or not.

There is one case that user can see Widevine prompt once again when that user installed widevine in their primary profile and not in non-primary profile and non-primary profile is firstly initialized profile. In this situation, firstly initialized profile is used for migration, user could see Widevine prompt just once more on macOS and Windows.

On Linux, prefs are recovered implicitly by (`SetWidevinePrefsAsInstalledState()`) in that case.

WDYT about modified logic?